consensus/src/dag/dag_store.rs

```diff
@@ -78,62 +96,154 @@ impl Dag {
+        round_ref[*index].as_ref()
+    }
+
+    fn get_round_iter(&self, round: Round) -> Option<impl Iterator<Item = &NodeStatus>> {
+        self.nodes_by_round
+            .get(&round)
+            .map(|round_ref| round_ref.iter().flatten())
+    }
+
+    pub fn get_node(&self, metadata: &NodeMetadata) -> Option<Arc<CertifiedNode>> {
+        self.get_node_ref_by_metadata(metadata)
+            .map(|node_status| node_status.as_node().clone())
+    }
+
+    pub fn get_node_by_round_author(
+        &self,
+        round: Round,
+        author: &Author,
+    ) -> Option<&Arc<CertifiedNode>> {
+        self.get_node_ref(round, author)
+            .map(|node_status| node_status.as_node())
+    }
+
+    // TODO: I think we can cache votes in the NodeStatus::Unordered
+    pub fn check_votes_for_node(
+        &self,
+        metadata: &NodeMetadata,
+        validator_verifier: &ValidatorVerifier,
+    ) -> bool {
+        self.get_round_iter(metadata.round() + 1)
+            .and_then(|next_round_iter| {
+                let votes = next_round_iter
+                    .filter(|node_status| {
+                        node_status
+                            .as_node()
+                            .parents()
+                            .iter()
+                            .any(|cert| cert.metadata() == metadata)
+                    })
+                    .map(|node_status| node_status.as_node().author());
+                Some(validator_verifier.check_voting_power(votes).is_ok())
+            })
+            .unwrap_or(false)
+    }
+
+    fn reachable_filter(start: HashValue) -> impl FnMut(&Arc<CertifiedNode>) -> bool {
+        let mut reachable = HashSet::from([start]);
+        move |node| {
+            for parent in node.parents() {
+                reachable.insert(*parent.metadata().digest());
+            }
+            reachable.contains(&node.digest())
+        }
+    }
+
+    pub fn reachable_mut<'a>(
+        &mut self,
+        from: &'a Arc<CertifiedNode>,
+        until: Option<Round>,
+    ) -> impl Iterator<Item = &mut NodeStatus> {
+        let until = until.unwrap_or(self.lowest_round());
+        let mut reachable_filter = Self::reachable_filter(from.digest());
+        self.nodes_by_round
+            .range_mut(until..=from.round())
+            .rev()
+            .flat_map(|(_, round_ref)| round_ref.iter_mut())
+            .flatten()
+            .filter(move |node_status| {
+                matches!(node_status, NodeStatus::Unordered(_))
+                    && reachable_filter(node_status.as_node())
+            })
+    }
+
+    pub fn reachable<'a>(
+        &self,
+        from: &'a Arc<CertifiedNode>,
+        until: Option<Round>,
+    ) -> impl Iterator<Item = &NodeStatus> {
+        let until = until.unwrap_or(self.lowest_round());
+        let mut reachable_filter = Self::reachable_filter(from.digest());
+        self.nodes_by_round
+            .range(until..=from.round())
```

Comment:
not a strong opinion, but I feel like we might visit fewer nodes if we iterate via the parents.

Reply:
yes, I thought about it, but given it needs to have 2f+1 parents, I don't think it matters that much. iterate through vector is probably faster than going through pointer lookups too